src/coreComponents/mesh/ParticleRegion.hpp

```diff
@@ -0,0 +1,124 @@
+/*
+ * ------------------------------------------------------------------------------------------------------------
+ * SPDX-License-Identifier: LGPL-2.1-only
+ *
+ * Copyright (c) 2018-2020 Lawrence Livermore National Security LLC
+ * Copyright (c) 2018-2020 The Board of Trustees of the Leland Stanford Junior University
+ * Copyright (c) 2018-2020 TotalEnergies
+ * Copyright (c) 2019-     GEOSX Contributors
+ * All rights reserved
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+
+/**
+ * @file ParticleRegion.hpp
+ *
+ */
+
+#ifndef GEOSX_MESH_PARTICLEREGION_HPP_
+#define GEOSX_MESH_PARTICLEREGION_HPP_
+
+#include "ElementRegionBase.hpp"
+
+namespace geosx
+{
+class EdgeManager;
+class EmbeddedSurfaceGenerator;
+
+/**
+ * @class ParticleRegion
+ *
+ * The ParticleRegion class contains the functionality to support the concept of a ParticleRegion in the element
+ * hierarchy. ParticleRegion derives from ElementRegionBase and has an entry in the ObjectManagerBase catalog.
+ *
+ *
+ */
+class ParticleRegion : public ElementRegionBase
+{
+public:
+
+  /**
+   * @name Constructor / Destructor
+   */
+  ///@{
+
+
+  /**
+   * @brief Constructor.
+   * @param name the name of the object in the data hierarchy.
+   * @param parent a pointer to the parent group in the data hierarchy.
+   */
+  ParticleRegion( string const & name, Group * const parent );
+
+  /**
+   * @brief Deleted default constructor.
+   */
+  ParticleRegion() = delete;
+
+  /**
+   * @brief Destructor.
+   */
+  virtual ~ParticleRegion() override;
+
+  /**
+   * @name Static factory catalog functions
+   */
+  ///@{
+
+  /**
+   * @brief The key name for the FaceElementRegion in the object catalog.
+   * @return A string containing the key name.
+   */
+  static const string catalogName()
+  { return "ParticleRegion"; }
+
+  /**
+   * @copydoc catalogName()
+   */
+  virtual const string getCatalogName() const override final
+  { return ParticleRegion::catalogName(); }
+
+  ///@}
+
+  /**
+   * @name Generation of the particle region
+   */
+  ///@{
+
+  /**
+   * @brief Add a cellBlockRegion name to the list.
+   * @param cellBlockName string containing the cell block region name.
+   */
+  void addCellBlockName( string const & cellBlockName )
+  {
+    m_cellBlockNames.emplace_back( cellBlockName );
+  }
+
+  virtual void generateMesh( Group & cellBlocks ) override;
+
+
+  ///@}
+
+  /**
+   * @brief A struct to serve as a container for variable strings and keys.
+   * @struct viewKeyStruct
+   */
+  struct viewKeyStruct : public ElementRegionBase::viewKeyStruct
+  {
+    /// @return String key for the cell block names
+    static constexpr char const * sourceCellBlockNamesString() {return "cellBlocks"; }
+  };
```

Comment:
Do you need this to be 
* `public`
* inherit from `ElementRegionBase::viewKeyStruct`

?

Reply:
I think that, in general, we do want the `viewKeyStruct` to be public. We don't have a getter for it and it is often used to get the keys inside solvers. This is a region though so I am not 100% sure it is actually useful.

Reply:
In this case, `viewKeyStruct::sourceCellBlockNamesString` was never used outside of the `ParticleRegion` class.
Exposing it `public`ly would in the end make the `cellBlocks` instances `public` too, while they are used as a `private` members in the present case.
If we have to keys in `viewKeyStruct`, one for a `public` attribute, the other for a `private` attribute, then the `private` attribute would become `public` even though it's not the desired pattern.

Having `viewKeyStruct ` `public` actually means that any registered `Wrapper`/`Group` could be accessed by anyone. This can beak encapsulation sometimes (like this time).